ttnn/cpp/ttnn/types_wrapper.cpp

```diff
@@ -0,0 +1,418 @@
+
+namespace ttnn::vector_wrapper
+{
+    tt::tt_metal::Shape to_shape(const std::vector<uint32_t>& shape_vector)
+    {
+        return tt::tt_metal::Shape(shape_vector);
+    }
+} // namespace ttnn::vector_wrapper
+
+namespace ttnn::tuple_wrapper {
+    std::optional<tt::tt_metal::ShardSpec> to_shard_spec(
+        const std::vector<std::array<uint32_t, 4>>& core_range_set,
+        const std::array<uint32_t, 2>& shard_shape,
+        const std::string& shard_orientation,
+        const bool& halo = false) {
+
+        if (core_range_set.size() == 0)
+        {
+            return std::nullopt;
+        }
+
+        std::set<CoreRange> core_ranges;
+        for (const auto& core_range : core_range_set)
+        {
+            core_ranges.insert(CoreRange{CoreCoord{core_range[0], core_range[1]}, CoreCoord{core_range[2], core_range[3]}});
+        }
+
+        return std::make_optional(tt::tt_metal::ShardSpec{
+            CoreRangeSet{core_ranges},
+            shard_shape,
+            str_wrapper::to_shard_orientation(shard_orientation).value(),
+            halo
+        });
+    }
+
+    std::optional<tt::tt_metal::ShardSpec> to_shard_spec(const mlir_interface::shard_spec_tuple& shard_spec_tuple) {
+        return to_shard_spec(
+            std::get<std::vector<std::array<uint32_t, 4>>>(shard_spec_tuple),
+            std::get<std::array<uint32_t, 2>>(shard_spec_tuple),
+            std::get<std::string>(shard_spec_tuple),
+            std::get<bool>(shard_spec_tuple));
+    }
+
+    std::optional<tt::tt_metal::MemoryConfig> to_memory_config(
+        const std::string& tensor_memory_layout,
+        const std::string& buffer_type,
+        const std::optional<mlir_interface::shard_spec_tuple>& shard_spec) {
+
+        auto deserialized_tensor_memory_layout = str_wrapper::to_tensor_memory_layout(tensor_memory_layout);
+        if (!deserialized_tensor_memory_layout.has_value()) {
+            return std::nullopt;
+        }
+
+        auto deserialized_buffer_type = str_wrapper::to_buffer_type(buffer_type);
+        if (!deserialized_buffer_type.has_value()) {
+            return std::nullopt;
+        }
+
+        auto deserialized_shard_spec = shard_spec.has_value() ? to_shard_spec(shard_spec.value()) : std::nullopt;
+        return std::make_optional(tt::tt_metal::MemoryConfig{
+            deserialized_tensor_memory_layout.value(),
+            deserialized_buffer_type.value(),
+            deserialized_shard_spec
+        });
+    }
+
+    std::optional<tt::tt_metal::MemoryConfig> to_memory_config(const mlir_interface::memory_config_tuple& memory_config_tuple) {
+        return to_memory_config(
+            std::get<0>(memory_config_tuple),
+            std::get<1>(memory_config_tuple),
+            std::get<std::optional<mlir_interface::shard_spec_tuple>>(memory_config_tuple));
+    }
+
+    // MatmulMultiCoreReuseProgramConfig wrapper
+    std::optional<ttnn::operations::matmul::MatmulMultiCoreReuseProgramConfig> to_program_config(
```

Comment:
nit: Should this maybe be named something like `to_multi_core_program_config`, to be in line with other the other two methods?